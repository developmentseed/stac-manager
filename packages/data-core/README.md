# @stac-manager/data-core

## Introduction
Core functionality of the form builder plugin system.

## Installation and Usage
See root README.md for instructions on how to install and use the project.
