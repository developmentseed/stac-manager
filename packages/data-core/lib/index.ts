export * from './plugin-utils/plugin';
export * from './plugin-utils/use-plugins-hook';
export * from './context/plugin-config';
export * from './config';
export * from './components/plugin-box';
export * from './components/error-box';
export * from './components/widget-renderer';

export * from './schema/types';
export * from './schema';
export * from './components/types';
