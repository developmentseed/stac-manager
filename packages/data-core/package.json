{
  "name": "@stac-manager/data-core",
  "description": "> TODO: description",
  "version": "1.0.0",
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.js",
  "types": "dist/index.d.ts",
  "source": "lib/index.ts",
  "sideEffects": false,
  "scripts": {
    "build": "rollup -c ../../rollup.config.mjs",
    "watch": "rollup -c ../../rollup.config.mjs -w",
    "clean": "rm -rf ./dist node_modules",
    "lint": "eslint ./lib --no-error-on-unmatched-pattern"
  },
  "engines": {
    "node": ">=20"
  },
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "directories": {
    "lib": "lib"
  },
  "dependencies": {
    "@chakra-ui/react": "^2.8.2",
    "@emotion/react": "^11.13.3",
    "@emotion/styled": "^11.13.0",
    "formik": "^2.4.6",
    "framer-motion": "^10.16.5",
    "lodash-es": "^4.17.21",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "devDependencies": {
    "@types/lodash-es": "^4.17.12",
    "@types/react": "^18.3.12",
    "@types/react-dom": "^18.3.1"
  },
  "peerDependencies": {
    "@chakra-ui/react": "^2.8.2",
    "@emotion/react": "^11.13.3",
    "@emotion/styled": "^11.13.0",
    "framer-motion": "^10.16.5",
    "formik": "^2.4.6",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  }
}
