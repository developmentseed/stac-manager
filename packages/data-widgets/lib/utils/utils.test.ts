import { mod, getArrayLabel, toNumber, castArray } from './index';
import { SchemaField } from '@stac-manager/data-core';

describe('mod', () => {
  it('should return the correct positive remainder', () => {
    expect(mod(5, 3)).toBe(2);
  });

  it('should handle negative dividends correctly', () => {
    expect(mod(-1, 3)).toBe(2);
  });

  it('should handle zero divisor correctly', () => {
    expect(mod(0, 3)).toBe(0);
  });
});

describe('getArrayLabel', () => {
  it('should return a label with a number suffix for string labels', () => {
    const field: SchemaField = { label: 'Label' } as any;
    const result = getArrayLabel(field, 9);

    expect(result).toEqual({
      label: 'Label',
      num: 10, // 1-based index
      formatted: expect.anything()
    });
    expect(result?.formatted).toMatchSnapshot();
  });

  it('should cycle through array labels based on index', () => {
    const field: SchemaField = { label: ['One', 'Two', 'Three'] } as any;
    expect(getArrayLabel(field, 0)?.label).toBe('One');
    expect(getArrayLabel(field, 3)?.label).toBe('One');
    expect(getArrayLabel(field, 4)?.label).toBe('Two');
  });

  it('should return null if label is null', () => {
    const field: SchemaField = { label: null } as any;
    expect(getArrayLabel(field, 0)).toBeNull();
  });

  it('should return Item if label is undefined', () => {
    const field: SchemaField = { label: undefined } as any;
    const result = getArrayLabel(field, 0);

    expect(result).toEqual({
      label: 'Item',
      num: 1,
      formatted: expect.anything()
    });
    expect(result?.formatted).toMatchSnapshot();
  });
});

describe('toNumber', () => {
  it('should convert valid numbers correctly', () => {
    expect(toNumber('42')).toBe(42);
    expect(toNumber(42)).toBe(42);
  });

  it('should return null for invalid numbers', () => {
    expect(toNumber('abc')).toBeNull();
    expect(toNumber(undefined)).toBeNull();
  });
});

describe('castArray', () => {
  it('should wrap non-array values in an array', () => {
    expect(castArray(42)).toEqual([42]);
    expect(castArray('test')).toEqual(['test']);
  });

  it('should return the same array if input is already an array', () => {
    expect(castArray([42])).toEqual([42]);
    expect(castArray(['test'])).toEqual(['test']);
  });
});
