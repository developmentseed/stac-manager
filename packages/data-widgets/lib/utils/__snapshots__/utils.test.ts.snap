// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getArrayLabel should return Item if label is undefined 1`] = `
<React.Fragment>
  <Text
    as="span"
  >
    Item
  </Text>
   
  <Text
    as="small"
  >
    01
  </Text>
</React.Fragment>
`;

exports[`getArrayLabel should return a label with a number suffix for string labels 1`] = `
<React.Fragment>
  <Text
    as="span"
  >
    Label
  </Text>
   
  <Text
    as="small"
  >
    10
  </Text>
</React.Fragment>
`;
